Core/PrimaryGenerator.cpp

```diff
@@ -264,34 +266,103 @@ void PrimaryGenerator::GenerateEvtLargeChamber(G4Event* event)
   }
   else if (evtFractions[0] + evtFractions[1] + evtFractions[2] + evtFractions[3] > random) // Direct 3G
   {
-    event->AddPrimaryVertex(GenerateThreeGammaVertex(vtxPosition, T0, 
+    event->AddPrimaryVertex(GenerateThreeGammaVertex(vtxPosition, T0,
                             material->GetLifetime(random - evtFractions[0] - evtFractions[1] - evtFractions[2], MaterialExtension::DecayChannel::Direct)));
   }
   else // oPs 3G
   {
-    event->AddPrimaryVertex( GenerateThreeGammaVertex(vtxPosition, T0, 
+    event->AddPrimaryVertex( GenerateThreeGammaVertex(vtxPosition, T0,
                             material->GetLifetime(random - evtFractions[1] - evtFractions[2] - evtFractions[3] - evtFractions[4], MaterialExtension::DecayChannel::Ortho3G)));
-  } 
+  }
 
   //! Add prompt gamma from sodium
-  G4ThreeVector promptVtxPosition =  VertexUniformInCylinder(0.2 * cm, 0.2 * cm) + chamberCenter;
-  event->AddPrimaryVertex(GeneratePromptGammaVertex( promptVtxPosition, 0.0f, 
-                            MaterialParameters::sodiumGammaTau, MaterialParameters::sodiumGammaEnergy));
+  G4ThreeVector promptVtxPosition = VertexUniformInCylinder(0.2 * cm, 0.2 * cm) + chamberCenter;
+  event->AddPrimaryVertex(GeneratePromptGammaVertex(
+    promptVtxPosition, 0.0f,
+    MaterialParameters::sodiumGammaTau,
+    MaterialParameters::sodiumGammaEnergy
+  ));
+}
+
+/**
+ * Generating vertex for cosmic radiation particle
+ * - choose randomly some point inside the detector cylinder,
+ * as cosmic particle has to fly through the detector anyway
+ * - choose particle momentum
+ * - choose muon charge with ratio of positive to negative muons
+ * in cosmic radiation at the sea level from https://arxiv.org/pdf/1005.5332.pdf
+ * - trace back the flight track of the particle to a point onto a cosmic roof,
+ * that is simulation worlds top dimension
+ * - add particle and vertex to an event with energy of 4 GeV, which is
+ * mean for muons at sea level, according to PDG (30.3.1)
+ * http://pdg.lbl.gov/2017/reviews/rpp2017-rev-cosmic-rays.pdf
+ */
+void PrimaryGenerator::GenerateCosmicVertex(G4Event* event)
+{
+  // Generating particle
+  G4ParticleDefinition* muonDefinition;
+  G4double muonFrac = DetectorConstants::muonChargeRatio/(1+DetectorConstants::muonChargeRatio);
+  if(muonFrac < G4UniformRand()){
+    muonDefinition = G4ParticleTable::GetParticleTable()->FindParticle("mu+");
+  } else {
+    muonDefinition = G4ParticleTable::GetParticleTable()->FindParticle("mu-");
+  }
+
+  // Generating angles
+  TF1 *cos2Func = new TF1("cos2", "pow(cos(x),2)", -twopi/4, twopi/4);
+  G4double theta = cos2Func->GetRandom();
+  G4double phi = G4UniformRand()*twopi;
+
+  // Cylinder of radius of third layer and half of sintillator dimension
+  G4ThreeVector posInDetector = VertexUniformInCylinder(
+    DetectorConstants::radius[2],
+    DetectorConstants::scinDim[2]/2
+  );
+
+  // projecting point from the cilinder onto "roof" of the simulation world
```

Comment:
small typo

Reply:
This could be changed into function I guess.

Reply:
applied